sources/board.cpp

```diff
@@ -2,19 +2,24 @@
 
 
 Board::Board(int width, int height) : nextId_{1}, killAll_{false} {
+  if (width <= 0 or height <= 0)
+      throw std::invalid_argument("Invalid board size. Should be grater than 0.");
   board_.resize(height);
   for (int i = 0; i < (int)board_.size(); ++i)
     board_[i].resize(width);
 }
 
 void Board::addWorm(WormType type, int x, int y) {
-  // TODO: Implement the body of the function. It should add a worm
-  // into the board_ and start a new thread which will invoke
-  // call operator - operator()(int) - defined in Worm class.
-  // Note: there are a few class members (like wormTypes_) that need to be
-  // updated as well to have full information about the worm that is being
-  // added. They are usually based on worm's id which should be uniquely given
-  // here.
+    int id = getNextId();
+
+    board_[x][y] = id;
+    wormTypes_[id] = type;
+    switch (type) {
+        case Lazy: worms_[id] = std::thread(LazyWorm(x, y, this), id);
```

Comment:
This looks a bit dangerous to me - it creates a thread and the assigns it to the map. These are two different operations, so I was worried about two threads instead of one. The documentation says that it's ok and the thread is aquired by the object on the left (place within `worms_`). http://www.cplusplus.com/reference/thread/thread/operator=/

I'd rather use `emplace`: https://en.cppreference.com/w/cpp/container/unordered_map/emplace
It is useful when you cannot copy or move some objects - it creates the object directly within the container. You don't need to create it and then push (like `push` `insert` etc.). Notice that some object may have this `operator=` deleted, so such approach will cause the compiler error.

Reply:
Thanks, I'll change this.